// formdata-node/lib/esm/FormData.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, privateMap, value) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to set private field on non-instance");
    }
    privateMap.set(receiver, value);
    return value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, privateMap) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to get private field on non-instance");
    }
    return privateMap.get(receiver);
};
var _stream, _content, _footer;
import { Readable } from "stream";
import { basename } from "path";
import { inspect } from "util";
import { File } from "./File.js";
import { fileFromPathSync } from "./fileFromPath.js";
import isFile from "./util/isFile.js";
import getLength from "./util/getLength.js";
import isPlainObject from "./util/isPlainObject.js";
import deprecateReadStream from "./util/deprecateReadStream.js";
import deprecateStreamProp from "./util/deprecateStreamProp.js";
import createBoundary from "./util/createBoundary.js";
import isReadStream from "./util/isReadStream.js";
import getFilename from "./util/getFilename.js";
const { isBuffer } = Buffer;
const DASHES = "-".repeat(2);
const CRLF = "\r\n";
const CRLF_BYTES_LENGTH = Buffer.byteLength(CRLF);
export class FormData {
    constructor(entries) {
        // TODO: Remove this along with FormData#stream getter in 4.x
        _stream.set(this, void 0);
        /**
         * Returns a boundary string
         */
        this.boundary = `NodeJSFormDataStreamBoundary${createBoundary()}`;
        /**
         * Stores internal data for every field
         */
        _content.set(this, new Map()
        /**
         * Returns field's footer
         */
        );
        /**
         * Returns field's footer
         */
        _footer.set(this, `${DASHES}${this.boundary}${DASHES}${CRLF.repeat(2)}`);
        __classPrivateFieldSet(this, _stream, Readable.from(this));
        if (entries) {
            entries.forEach(({ name, value, filename, options }) => this.append(name, value, filename, options));
        }
    }
    /**
     * Returns headers for multipart/form-data
     */
    get headers() {
        return {
            "Content-Type": `multipart/form-data; boundary=${this.boundary}`
        };
    }
    /**
     * Returns internal readable stream, allowing to read the FormData content
     *
     * @deprecated Use `Readable.from(formData)` to create a stream from `FormData` instance.
     */
    get stream() {
        return __classPrivateFieldGet(this, _stream);
    }
    _getHeader(name, value) {
        let header = "";
        header += `${DASHES}${this.boundary}${CRLF}`;
        header += `Content-Disposition: form-data; name="${name}"`;
        if (isFile(value)) {
            header += `; filename="${value.name}"${CRLF}`;
            header += `Content-Type: ${value.type || "application/octet-stream"}`;
        }
        return `${header}${CRLF.repeat(2)}`;
    }
    async *_getField() {
        // Note to switch back to reading from this.#content if any extra logic will be necessary in a future, because the public FormData API returns values only as `string | File`
        for (const [name, value] of this) {
            // Set field's header
            yield this._getHeader(name, value);
            if (isFile(value)) {
                yield* value.stream();
            }
            else {
                yield value;
            }
            // Add trailing carriage
            yield CRLF;
        }
        // Add a footer when all fields ended
        yield __classPrivateFieldGet(this, _footer);
    }
    _setField({ name, value, append, filenameOrOptions, options, argsLength }) {
        const fieldName = String(name);
        const methodName = append ? "append" : "set";
        let filename;
        if (isPlainObject(filenameOrOptions)) {
            [options, filename] = [filenameOrOptions, undefined];
        }
        else {
            filename = filenameOrOptions;
        }
        // FormData required at least 2 arguments to be set.
        if (argsLength < 2) {
            throw new TypeError(`Failed to execute '${methodName}' on 'FormData': `
                + `2 arguments required, but only ${argsLength} present.`);
        }
        // Get a filename from either an argument or options
        filename || (filename = options === null || options === void 0 ? void 0 : options.filename);
        // If a value is a file-like object, then get and normalize the filename
        if (isFile(value) || isReadStream(value) || isBuffer(value)) {
            // Note that the user-defined filename has higher precedence
            filename = basename(filename || getFilename(value));
        }
        else if (filename) { // If a value is not a file-like, but the filename is present, then throw the error
            throw new TypeError(`Failed to execute '${methodName}' on 'FormData': `
                + "parameter 2 is not one of the following types: "
                + "ReadStream | Buffer | File | Blob");
        }
        // Normalize field's value
        if (isReadStream(value)) {
            // TODO: Remove ReadStream support in favour of fileFromPath and fileFromPathSync
            value = fileFromPathSync(String(value.path), filename, options);
        }
        else if (isBuffer(value)) {
            // TODO: Remove Buffer in a field's value support in favour of Blob.
            value = new File([value], filename, options);
        }
        else if (isFile(value)) {
            value = new File([value], filename, {
                ...options,
                // Take params from the previous File or Blob instance
                // But keep user-defined options higher percidence
                type: (options === null || options === void 0 ? void 0 : options.type) || value.type,
                lastModified: (options === null || options === void 0 ? void 0 : options.lastModified) || value.lastModified
            });
        }
        else {
            // A non-file fields must be converted to string
            value = String(value);
        }
        const field = __classPrivateFieldGet(this, _content).get(fieldName);
        if (!field) {
            return void __classPrivateFieldGet(this, _content).set(fieldName, {
                append,
                values: [value]
            });
        }
        // Replace a value of the existing field if "set" called
        if (!append) {
            return void __classPrivateFieldGet(this, _content).set(fieldName, {
                append,
                values: [value]
            });
        }
        // Do nothing if the field has been created from .set()
        if (!field.append) {
            return undefined;
        }
        // Append a new value to the existing field
        field.values.push(value);
        __classPrivateFieldGet(this, _content).set(fieldName, field);
    }
    /**
     * Returns computed length of the FormData content.
     */
    getComputedLength() {
        let length = 0;
        for (const [name, value] of this) {
            length += Buffer.byteLength(this._getHeader(name, value));
            length += getLength(value) + CRLF_BYTES_LENGTH;
        }
        return length + Buffer.byteLength(__classPrivateFieldGet(this, _footer));
    }
    append(name, value, filenameOrOptions, options) {
        return this._setField({
            name,
            value,
            filenameOrOptions,
            options,
            append: true,
            argsLength: arguments.length
        });
    }
    set(name, value, filenameOrOptions, options) {
        return this._setField({
            name,
            value,
            filenameOrOptions,
            options,
            append: false,
            argsLength: arguments.length
        });
    }
    /**
     * Returns the first value associated with the given name.
     * Buffer and Readable values will be returned as-is.
     *
     * @param {string} name A name of the value you want to retrieve.
     */
    get(name) {
        name = String(name);
        const field = __classPrivateFieldGet(this, _content).get(name);
        if (!field) {
            return null;
        }
        return field.values[0];
    }
    /**
     * Returns all the values associated with
     * a given key from within a FormData object.
     *
     * @param {string} name A name of the value you want to retrieve.
     */
    getAll(name) {
        name = String(name);
        const field = __classPrivateFieldGet(this, _content).get(name);
        if (!field) {
            return [];
        }
        return [...field.values];
    }
    /**
     * Check if a field with the given name exists inside FormData.
     *
     * @param name A name of the field you want to test for.
     *
     * @return
     */
    has(name) {
        return __classPrivateFieldGet(this, _content).has(String(name));
    }
    /**
     * Deletes a key and its value(s) from a FormData object.
     *
     * @param name The name of the key you want to delete.
     */
    delete(name) {
        return void __classPrivateFieldGet(this, _content).delete(String(name));
    }
    /**
     * Returns an [`iterator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols) allowing to go through the **FormData** keys
     */
    *keys() {
        for (const key of __classPrivateFieldGet(this, _content).keys()) {
            yield key;
        }
    }
    /**
     * Returns an [`iterator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols) allowing to go through the **FormData** key/value pairs
     */
    *entries() {
        for (const name of this.keys()) {
            const values = this.getAll(name);
            // Yield each value of a field, like browser-side FormData does.
            for (const value of values) {
                yield [name, value];
            }
        }
    }
    /**
     * Returns an [`iterator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols) allowing to go through the **FormData** values
     */
    *values() {
        for (const [, value] of this) {
            yield value;
        }
    }
    /**
     * An alias for FormData#entries()
     */
    [(_stream = new WeakMap(), _content = new WeakMap(), _footer = new WeakMap(), Symbol.iterator)]() {
        return this.entries();
    }
    /**
     * Executes given callback function for each field of the FormData instance
     */
    forEach(fn, ctx) {
        for (const [name, value] of this) {
            fn.call(ctx, value, name, this);
        }
    }
    get [Symbol.toStringTag]() {
        return "FormData";
    }
    [inspect.custom]() {
        return this[Symbol.toStringTag];
    }
    /**
     * Returns an async iterator allowing to read form-data body using **for-await-of** syntax.
     * Read the [`async iteration proposal`](https://github.com/tc39/proposal-async-iteration) to get more info about async iterators.
     */
    async *[Symbol.asyncIterator]() {
        for await (const ch of this._getField()) {
            yield isBuffer(ch) ? ch : Buffer.from(String(ch));
        }
    }
}
__decorate([
    deprecateStreamProp
], FormData.prototype, "stream", null);
__decorate([
    deprecateReadStream
], FormData.prototype, "append", null);
__decorate([
    deprecateReadStream
], FormData.prototype, "set", null);
